### PR TITLE
Feature/preprocessing

### LINKED ISSUE #3: Bug - SSLCertVerificationError
---------------------------------------------------------------------------
SSLCertVerificationError                  Traceback (most recent call last)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in do_open(self, http_class, req, **http_conn_args)
   1316                 h.request(req.get_method(), req.selector, req.data, headers,
-> 1317                           encode_chunked=req.has_header('Transfer-encoding'))
   1318             except OSError as err: # timeout error

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py in request(self, method, url, body, headers, encode_chunked)
   1228         """Send a complete request to the server."""
-> 1229         self._send_request(method, url, body, headers, encode_chunked)
   1230 

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py in _send_request(self, method, url, body, headers, encode_chunked)
   1274             body = _encode(body, 'body')
-> 1275         self.endheaders(body, encode_chunked=encode_chunked)
   1276 

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py in endheaders(self, message_body, encode_chunked)
   1223             raise CannotSendHeader()
-> 1224         self._send_output(message_body, encode_chunked=encode_chunked)
   1225 

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py in _send_output(self, message_body, encode_chunked)
   1015         del self._buffer[:]
-> 1016         self.send(msg)
   1017 

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py in send(self, data)
    955             if self.auto_open:
--> 956                 self.connect()
    957             else:

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py in connect(self)
   1391             self.sock = self._context.wrap_socket(self.sock,
-> 1392                                                   server_hostname=server_hostname)
   1393 

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py in wrap_socket(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)
    411             context=self,
--> 412             session=session
    413         )

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py in _create(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)
    852                         raise ValueError("do_handshake_on_connect should not be specified for non-blocking sockets")
--> 853                     self.do_handshake()
    854             except (OSError, ValueError):

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py in do_handshake(self, block)
   1116                 self.settimeout(None)
-> 1117             self._sslobj.do_handshake()
   1118         finally:

SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)

During handling of the above exception, another exception occurred:

URLError                                  Traceback (most recent call last)
<ipython-input-14-e79093d5829c> in <module>
----> 1 stations_df = pd.read_csv('https://mannheim.opendatasoft.com/explore/dataset/free_bike_status/download/?format=csv&timezone=Europe/Berlin&lang=de&use_labels_for_header=true&csv_separator=%3B', delimiter=';')
      2 stations_df.head()

~/PycharmProjects/PDS_Project/venv/lib/python3.7/site-packages/pandas/io/parsers.py in parser_f(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)
    674         )
    675 
--> 676         return _read(filepath_or_buffer, kwds)
    677 
    678     parser_f.__name__ = name

~/PycharmProjects/PDS_Project/venv/lib/python3.7/site-packages/pandas/io/parsers.py in _read(filepath_or_buffer, kwds)
    429     # See https://github.com/python/mypy/issues/1297
    430     fp_or_buf, _, compression, should_close = get_filepath_or_buffer(
--> 431         filepath_or_buffer, encoding, compression
    432     )
    433     kwds["compression"] = compression

~/PycharmProjects/PDS_Project/venv/lib/python3.7/site-packages/pandas/io/common.py in get_filepath_or_buffer(filepath_or_buffer, encoding, compression, mode)
    170 
    171     if isinstance(filepath_or_buffer, str) and is_url(filepath_or_buffer):
--> 172         req = urlopen(filepath_or_buffer)
    173         content_encoding = req.headers.get("Content-Encoding", None)
    174         if content_encoding == "gzip":

~/PycharmProjects/PDS_Project/venv/lib/python3.7/site-packages/pandas/io/common.py in urlopen(*args, **kwargs)
    139     import urllib.request
    140 
--> 141     return urllib.request.urlopen(*args, **kwargs)
    142 
    143 

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in urlopen(url, data, timeout, cafile, capath, cadefault, context)
    220     else:
    221         opener = _opener
--> 222     return opener.open(url, data, timeout)
    223 
    224 def install_opener(opener):

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in open(self, fullurl, data, timeout)
    523             req = meth(req)
    524 
--> 525         response = self._open(req, data)
    526 
    527         # post-process response

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in _open(self, req, data)
    541         protocol = req.type
    542         result = self._call_chain(self.handle_open, protocol, protocol +
--> 543                                   '_open', req)
    544         if result:
    545             return result

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in _call_chain(self, chain, kind, meth_name, *args)
    501         for handler in handlers:
    502             func = getattr(handler, meth_name)
--> 503             result = func(*args)
    504             if result is not None:
    505                 return result

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in https_open(self, req)
   1358         def https_open(self, req):
   1359             return self.do_open(http.client.HTTPSConnection, req,
-> 1360                 context=self._context, check_hostname=self._check_hostname)
   1361 
   1362         https_request = AbstractHTTPHandler.do_request_

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py in do_open(self, http_class, req, **http_conn_args)
   1317                           encode_chunked=req.has_header('Transfer-encoding'))
   1318             except OSError as err: # timeout error
-> 1319                 raise URLError(err)
   1320             r = h.getresponse()
   1321         except:

URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)>

### PR DESCRIPTION
**Pull request by phipag**
I finished the first part of the project. Please leave me some feedback.

### CODE REVIEW

**Comment by LHumpe** [comment]
Please omit printing of first line in Jupyter notebook

**Comment by LHumpe** [comment]
Please comment all Classes and Functions properly:
Example:

```
class Transformer(AbstractValidator):
    """
    A class that inherits from AbstractValidator handling xyz
    """
    __preprocessor: Preprocessor
    __gdf: gpd.GeoDataFrame = None

    @property
    def gdf(self) -> gpd.GeoDataFrame:
        """
        Does not need an import param
        :return: a Geodataframe containing information xyz
        """
        if self.__gdf is None:
            raise UserWarning('Data frame is not initialized.')
        return self.__gdf

    def __init__(self, preprocessor: Preprocessor):
        """
        Function for initialising the preprocessor class
        :param preprocessor: the class preprocessor
        """
        try:
            if preprocessor.validate():
                self.__preprocessor = preprocessor
        except ValueError:
            raise ValueError('Preprocessor validation failed. Please make sure that the Preprocessor was successful.')```

**Review by LHumpe — CHANGES_REQUESTED** [review]
Please comment all Classes and Functions properly:
Example:

```
class Transformer(AbstractValidator):
    """
    A class that inherits from AbstractValidator handling xyz
    """
    __preprocessor: Preprocessor
    __gdf: gpd.GeoDataFrame = None

    @property
    def gdf(self) -> gpd.GeoDataFrame:
        """
        Does not need an import param
        :return: a Geodataframe containing information xyz
        """
        if self.__gdf is None:
            raise UserWarning('Data frame is not initialized.')
        return self.__gdf

    def __init__(self, preprocessor: Preprocessor):
        """
        Function for initialising the preprocessor class
        :param preprocessor: the class preprocessor
        """
        try:
            if preprocessor.validate():
                self.__preprocessor = preprocessor
        except ValueError:
            raise ValueError('Preprocessor validation failed. Please make sure that the Preprocessor was successful.')```

File: nextbike/preprocessing/Transformer.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        return self.__gdf
+
+    def __init__(self, preprocessor: Preprocessor):
+        try:
+            if preprocessor.validate():
+                self.__preprocessor = preprocessor
+        except ValueError:
+            raise ValueError('Preprocessor validation failed. Please make sure that the Preprocessor was successful.')
+
+    def transform(self, validate=False) -> None:
```
Comment (LHumpe) [inline nextbike/preprocessing/Transformer.py]: Add dummy variable is_station utilising p_place_type

Keep station names for start and end 

Duration as float or int

Comment (LHumpe) [reply nextbike/preprocessing/Transformer.py]: Add columns month_of_year, day_of_week(weekday), hour_of_day, Season

**Comment by phipag** [comment]
> Please omit printing of first line in Jupyter notebook

I don't think this is a good idea because the user wants to see if the package could be installed correctly.

**Comment by phipag** [comment]
> Please comment all Classes and Functions properly:
> Example:
> 
> ```
> class Transformer(AbstractValidator):
>     """
>     A class that inherits from AbstractValidator handling xyz
>     """
>     __preprocessor: Preprocessor
>     __gdf: gpd.GeoDataFrame = None
> 
>     @property
>     def gdf(self) -> gpd.GeoDataFrame:
>         """
>         Does not need an import param
>         :return: a Geodataframe containing information xyz
>         """
>         if self.__gdf is None:
>             raise UserWarning('Data frame is not initialized.')
>         return self.__gdf
> 
>     def __init__(self, preprocessor: Preprocessor):
>         """
>         Function for initialising the preprocessor class
>         :param preprocessor: the class preprocessor
>         """
>         try:
>             if preprocessor.validate():
>                 self.__preprocessor = preprocessor
>         except ValueError:
>             raise ValueError('Preprocessor validation failed. Please make sure that the Preprocessor was successful.')```
> ```

Done by a3fd9c90

File: nextbike/preprocessing/Transformer.py (hunk 1 above)
Comment (phipag) [reply nextbike/preprocessing/Transformer.py]: Done by 4bc6453f

**Review by LHumpe — COMMENTED** [review]
Well done Philipp! Thank you

**Review by LHumpe — APPROVED** [review]

**Pull request merged**